backends/audiocd/win32/win32-audiocd.cpp

```diff
@@ -0,0 +1,388 @@
+	bool openCD(const Common::String &drive);
+
+private:
+	bool loadTOC();
+
+	typedef Common::Array<char> DriveList;
+	DriveList detectDrives();
+	bool tryAddDrive(char drive, DriveList &drives);
+
+	HANDLE _driveHandle;
+	int _firstTrack, _lastTrack;
+	Common::Array<TRACK_DATA> _tocEntries;
+};
+
+Win32AudioCDManager::Win32AudioCDManager() {
+	_driveHandle = INVALID_HANDLE_VALUE;
+	_firstTrack = _lastTrack = 0;
+}
+
+Win32AudioCDManager::~Win32AudioCDManager() {
+	close();
+}
+
+bool Win32AudioCDManager::open() {
+	close();
+
+	if (openRealCD())
+		return true;
+
+	return DefaultAudioCDManager::open();
+}
+
+bool Win32AudioCDManager::openCD(int drive) {
+	// Fetch the drive list
+	DriveList drives = detectDrives();
+	if (drive >= (int)drives.size())
+		return false;
+
+	debug(1, "Opening CD drive %c:\\", drives[drive]);
+
+	// Construct the drive path and try to open it
+	Common::String drivePath = Common::String::format("\\\\.\\%c:", drives[drive]);
+	_driveHandle = CreateFileA(drivePath.c_str(), GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, 0, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, 0);
+	if (_driveHandle == INVALID_HANDLE_VALUE) {
+		warning("Failed to open drive %c:\\, error %d", drives[drive], (int)GetLastError());
+		return false;
+	}
+
+	if (!loadTOC()) {
+		close();
+		return false;
+	}
+
+	return true;
+}
+
+bool Win32AudioCDManager::openCD(const Common::String &drive) {
+	// Just some bounds checking
+	if (drive.empty() || drive.size() > 3)
+		return false;
+
+	if (!Common::isAlpha(drive[0]) || drive[1] != ':')
+		return false;
+
+	if (drive[2] != 0 && drive[2] != '\\')
+		return false;
+
+	DriveList drives;
+	if (!tryAddDrive(toupper(drive[0]), drives))
+		return false;
+
+	// Construct the drive path and try to open it
+	Common::String drivePath = Common::String::format("\\\\.\\%c:", drives[0]);
+	_driveHandle = CreateFileA(drivePath.c_str(), GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE, 0, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, 0);
+	if (_driveHandle == INVALID_HANDLE_VALUE) {
+		warning("Failed to open drive %c:\\, error %d", drives[0], (int)GetLastError());
+		return false;
+	}
+
+	if (!loadTOC()) {
+		close();
+		return false;
+	}
+
+	return true;
+}
+
+void Win32AudioCDManager::close() {
+	DefaultAudioCDManager::close();
+
+	if (_driveHandle != INVALID_HANDLE_VALUE) {
+		CloseHandle(_driveHandle);
+		_driveHandle = INVALID_HANDLE_VALUE;
+	}
+
+	_firstTrack = _lastTrack = 0;
+	_tocEntries.clear();
+}
+
+bool Win32AudioCDManager::play(int track, int numLoops, int startFrame, int duration, bool onlyEmulate) {
+	// Prefer emulation
+	if (DefaultAudioCDManager::play(track, numLoops, startFrame, duration, onlyEmulate))
+		return true;
+
+	// If we're set to only emulate, or have no CD drive, return here
+	if (onlyEmulate || _driveHandle == INVALID_HANDLE_VALUE)
+		return false;
+
+	// HACK: For now, just assume that track number is right
+	// That only works because ScummVM uses the wrong track number anyway
```

Comment:
There seems to be a check on the track number below, is this comment still accurate?